DataFormats/MemoryResources/include/MemoryResources/MemoryResources.h

```diff
@@ -0,0 +1,274 @@
+                                        typename ContainerT::allocator_type>::value == true,
+                        FairMQMessagePtr>::type
+  getMessage(ContainerT&& container_, o2::memoryResources::FairMQMemoryResource* targetResource = nullptr)
+{
+  auto container = std::move(container_);
+  auto alloc = container.get_allocator();
+
+  auto resource = dynamic_cast<o2::memoryResources::FairMQMemoryResource*>(alloc.resource());
+  if (!resource && !targetResource) {
+    throw std::runtime_error("Neither the container or target resource specified");
+  }
+  size_t containerSizeBytes = container.size() * sizeof(typename ContainerT::value_type);
+  if ((!targetResource && resource) || (resource && targetResource && resource->is_equal(*targetResource))) {
+    auto message = resource->getMessage(static_cast<void*>(container.data()));
+    message->SetUsedSize(containerSizeBytes);
+    return std::move(message);
+  } else {
+    auto message = targetResource->getTransportFactory()->CreateMessage(containerSizeBytes);
+    std::memcpy(static_cast<byte*>(message->GetData()), container.data(), containerSizeBytes);
+    return std::move(message);
+  }
+};
+
+//__________________________________________________________________________________________________
+// This returns a unique_ptr of const vector, does not allow modifications at the cost of pointer
+// semantics for access.
+// use auto or decltype to catch the return value (or use span)
+template <typename ElemT>
+auto adoptVector(size_t nelem, FairMQMessage* message)
+{
+  using namespace o2::memoryResources;
+  using DataType = std::vector<ElemT, FastSpectatorAllocator>;
+
+  struct doubleDeleter {
+    // kids: don't do this at home! (but here it's OK)
+    // this stateful deleter allows a single unique_ptr to manage 2 resources at the same time.
+    std::unique_ptr<SpectatorMessageResource> extra;
+    void operator()(const DataType* ptr) { delete ptr; }
+  };
+
+  using OutputType = std::unique_ptr<const DataType, doubleDeleter>;
+
+  auto resource = std::make_unique<SpectatorMessageResource>(message);
+  auto output = new DataType(nelem, FastSpectatorAllocator{ resource.get() });
+  return OutputType(output, doubleDeleter{ std::move(resource) });
+}
+
+namespace internal
+{
+//__________________________________________________________________________________________________
+/// A (thread local) singleton placeholder for the channel allocators. There will normally be 1-2 elements in the map.
+// Ideally the transport class itself would hold (or be) the allocator, if that ever happens, this can go away.
+class TransportAllocatorMap
+{
+ public:
+  static TransportAllocatorMap& Instance()
+  {
+    static TransportAllocatorMap S;
+    return S;
+  }
+  ChannelResource* operator[](const FairMQTransportFactory* factory) { return std::addressof(map.emplace(factory, factory).first->second); }
```

Comment:
doesn't the pointer change if more elements are added to the map?

Reply:
not that I know. I think only erase would invalidate the pointer, documentation does not say anything about invalidation on emplace/insert. Nodes are allocated for good and never moved/copied since (unordered_)map is not a contiguous thing...

Reply:
actually the spec says emplace does not invalidate anything